File: collectors/logs.go
package collectors

import (
	"bufio"
	"encoding/json"
	"os"
	"os/exec"
	"regexp"
	"sort"
	"strings"

	"ubuntu-state/models"
)

// journalEntry represents a journalctl JSON output entry
type journalEntry struct {
	Priority string `json:"PRIORITY"`
	Message  string `json:"MESSAGE"`
}

// CollectLogInfo analyzes system logs for the last 24 hours
func CollectLogInfo() models.LogInfo {
	info := models.LogInfo{
		Period: "24h",
		Stats: models.LogStats{
			TopErrors: []models.ErrorPattern{},
		},
	}

	// Check if journalctl is available
	if _, err := exec.LookPath("journalctl"); err != nil {
		return info
	}
	info.Available = true

	// Get log statistics
	info.Stats = analyzeJournalLogs()

	// Check for OOM events
	info.Stats.OOMEvents = countOOMEvents()

	// Check for kernel panics
	info.Stats.KernelPanics = countKernelPanics()

	// Check for segfaults
	info.Stats.Segfaults = countSegfaults()

	return info
}

func analyzeJournalLogs() models.LogStats {
	stats := models.LogStats{
		TopErrors: []models.ErrorPattern{},
	}

	// Get errors and above (priority 0-3)
	cmd := exec.Command("journalctl",
		"--since", "24 hours ago",
		"-p", "err..emerg",
		"--no-pager",
		"-o", "json")
	output, err := cmd.Output()
	if err != nil {
		return stats
	}

	// Parse JSON entries and count by priority
	errorPatterns := make(map[string]int)

	lines := strings.Split(string(output), "\n")
	for _, line := range lines {
		if line == "" {
			continue
		}

		var entry journalEntry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			continue
		}

		// Count by priority
		// 0: emerg, 1: alert, 2: crit, 3: err, 4: warning, 5: notice, 6: info, 7: debug
		switch entry.Priority {
		case "0", "1", "2":
			stats.CriticalCount++
		case "3":
			stats.ErrorCount++
		case "4":
			stats.WarningCount++
		}

		// Extract error pattern (simplify message for grouping)
		if entry.Message != "" {
			pattern := simplifyErrorMessage(entry.Message)
			if pattern != "" {
				errorPatterns[pattern]++
			}
		}
	}

	// Get top 5 error patterns
	stats.TopErrors = getTopPatterns(errorPatterns, 5)

	return stats
}

func simplifyErrorMessage(msg string) string {
	// Remove timestamps, PIDs, memory addresses, etc.
	msg = strings.TrimSpace(msg)

	// Remove hex addresses
	re := regexp.MustCompile(`0x[0-9a-fA-F]+`)
	msg = re.ReplaceAllString(msg, "0x...")

	// Remove numbers that look like PIDs or ports
	re = regexp.MustCompile(`\b\d{4,}\b`)
	msg = re.ReplaceAllString(msg, "...")

	// Truncate long messages
	if len(msg) > 80 {
		msg = msg[:80] + "..."
	}

	return msg
}

func getTopPatterns(patterns map[string]int, limit int) []models.ErrorPattern {
	var result []models.ErrorPattern

	for pattern, count := range patterns {
		result = append(result, models.ErrorPattern{
			Pattern: pattern,
			Count:   count,
		})
	}

	// Sort by count descending
	sort.Slice(result, func(i, j int) bool {
		return result[i].Count > result[j].Count
	})

	// Limit results
	if len(result) > limit {
		result = result[:limit]
	}

	return result
}

func countOOMEvents() int {
	cmd := exec.Command("journalctl",
		"--since", "24 hours ago",
		"-k",
		"--no-pager",
		"--grep", "Out of memory|oom-kill|oom_reaper")
	output, err := cmd.Output()
	if err != nil {
		// Command returns non-zero if no matches
		return 0
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	count := 0
	for _, line := range lines {
		if line != "" {
			count++
		}
	}
	return count
}

func countKernelPanics() int {
	// Check journalctl for kernel panics
	cmd := exec.Command("journalctl",
		"--since", "24 hours ago",
		"-k",
		"--no-pager",
		"--grep", "Kernel panic")
	output, _ := cmd.Output()

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	count := 0
	for _, line := range lines {
		if line != "" {
			count++
		}
	}

	// Also check /var/log/kern.log if it exists
	if f, err := os.Open("/var/log/kern.log"); err == nil {
		defer f.Close()
		scanner := bufio.NewScanner(f)
		for scanner.Scan() {
			if strings.Contains(strings.ToLower(scanner.Text()), "kernel panic") {
				count++
			}
		}
	}

	return count
}

func countSegfaults() int {
	cmd := exec.Command("journalctl",
		"--since", "24 hours ago",
		"-k",
		"--no-pager",
		"--grep", "segfault")
	output, err := cmd.Output()
	if err != nil {
		return 0
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	count := 0
	for _, line := range lines {
		if line != "" {
			count++
		}
	}
	return count
}

File: zig/src/collectors/logs.zig
const std = @import("std");
const utils = @import("../utils.zig");
const report = @import("../models/report.zig");

pub fn collectLogInfo(allocator: std.mem.Allocator) ?report.LogInfo {
    if (!utils.commandExists("journalctl")) {
        return report.LogInfo{
            .available = false,
            .stats = report.LogStats{},
        };
    }

    var stats = report.LogStats{};

    // Get errors and above with JSON output
    const output = utils.runCommand(allocator, &[_][]const u8{
        "journalctl",
        "--since",
        "24 hours ago",
        "-p",
        "err..emerg",
        "--no-pager",
        "-o",
        "json",
    }) catch "";

    var error_patterns = std.ArrayList(report.ErrorPattern).empty;
    var pattern_counts = std.StringHashMap(i32).init(allocator);
    defer pattern_counts.deinit();

    if (output.len > 0) {
        var lines = std.mem.splitScalar(u8, output, '\n');
        while (lines.next()) |line| {
            if (line.len == 0) continue;

            // Parse JSON entry
            const parsed = std.json.parseFromSlice(std.json.Value, allocator, line, .{}) catch continue;
            defer parsed.deinit();

            const obj = parsed.value.object;

            // Get priority
            if (obj.get("PRIORITY")) |priority_val| {
                const priority_str = switch (priority_val) {
                    .string => |s| s,
                    .integer => |i| blk: {
                        var buf: [8]u8 = undefined;
                        break :blk std.fmt.bufPrint(&buf, "{d}", .{i}) catch "3";
                    },
                    else => "3",
                };
                // Priority: 0=emerg, 1=alert, 2=crit, 3=err, 4=warning
                if (std.mem.eql(u8, priority_str, "0") or
                    std.mem.eql(u8, priority_str, "1") or
                    std.mem.eql(u8, priority_str, "2"))
                {
                    stats.critical_count += 1;
                } else if (std.mem.eql(u8, priority_str, "3")) {
                    stats.error_count += 1;
                } else if (std.mem.eql(u8, priority_str, "4")) {
                    stats.warning_count += 1;
                }
            }

            // Get message for pattern extraction
            if (obj.get("MESSAGE")) |msg_val| {
                if (msg_val == .string) {
                    const msg = msg_val.string;
                    const pattern = simplifyLogPattern(allocator, msg);
                    if (pattern.len > 0 and pattern.len < 200) {
                        const result = pattern_counts.getOrPut(pattern) catch continue;
                        if (result.found_existing) {
                            result.value_ptr.* += 1;
                        } else {
                            result.value_ptr.* = 1;
                        }
                    }
                }
            }
        }
    }

    // Count OOM events with separate grep
    stats.oom_events = countGrepMatches(allocator, "Out of memory|oom-kill|oom_reaper");

    // Count kernel panics
    stats.kernel_panics = countGrepMatches(allocator, "Kernel panic");

    // Count segfaults
    stats.segfaults = countGrepMatches(allocator, "segfault");

    // Convert to top errors
    var pattern_list = std.ArrayList(PatternCount).empty;
    defer pattern_list.deinit(allocator);

    var it = pattern_counts.iterator();
    while (it.next()) |entry| {
        pattern_list.append(allocator, .{
            .pattern = entry.key_ptr.*,
            .count = entry.value_ptr.*,
        }) catch continue;
    }

    std.mem.sort(PatternCount, pattern_list.items, {}, struct {
        fn lessThan(_: void, a: PatternCount, b: PatternCount) bool {
            return a.count > b.count;
        }
    }.lessThan);

    const max_patterns = @min(pattern_list.items.len, 5);
    for (pattern_list.items[0..max_patterns]) |pc| {
        error_patterns.append(allocator, report.ErrorPattern{
            .pattern = allocator.dupe(u8, pc.pattern) catch continue,
            .count = pc.count,
        }) catch continue;
    }

    stats.top_errors = error_patterns.toOwnedSlice(allocator) catch null;

    return report.LogInfo{
        .available = true,
        .period = "24h",
        .stats = stats,
    };
}

fn countGrepMatches(allocator: std.mem.Allocator, grep_pattern: []const u8) i32 {
    const output = utils.runCommand(allocator, &[_][]const u8{
        "journalctl",
        "--since",
        "24 hours ago",
        "-k",
        "--no-pager",
        "--grep",
        grep_pattern,
    }) catch return 0;

    var count: i32 = 0;
    var lines = std.mem.splitScalar(u8, output, '\n');
    while (lines.next()) |line| {
        if (line.len > 0) count += 1;
    }
    return count;
}

const PatternCount = struct {
    pattern: []const u8,
    count: i32,
};

fn simplifyLogPattern(allocator: std.mem.Allocator, msg: []const u8) []const u8 {
    if (msg.len == 0) return "";

    // Simplify: replace hex addresses and long numbers
    var result = std.ArrayList(u8).empty;
    var i: usize = 0;
    while (i < msg.len) {
        // Check for hex address (0x...)
        if (i + 2 < msg.len and msg[i] == '0' and msg[i + 1] == 'x') {
            result.appendSlice(allocator, "0x...") catch return msg[0..@min(msg.len, 80)];
            i += 2;
            while (i < msg.len and isHexChar(msg[i])) : (i += 1) {}
            continue;
        }

        // Check for long numbers (4+ digits)
        if (isDigit(msg[i])) {
            var num_len: usize = 0;
            const start = i;
            while (i < msg.len and isDigit(msg[i])) : (i += 1) {
                num_len += 1;
            }
            if (num_len >= 4) {
                result.appendSlice(allocator, "...") catch return msg[0..@min(msg.len, 80)];
            } else {
                result.appendSlice(allocator, msg[start..i]) catch return msg[0..@min(msg.len, 80)];
            }
            continue;
        }

        result.append(allocator, msg[i]) catch return msg[0..@min(msg.len, 80)];
        i += 1;
    }

    const final = result.toOwnedSlice(allocator) catch return msg[0..@min(msg.len, 80)];
    if (final.len > 80) {
        return allocator.dupe(u8, final[0..77] ++ "...") catch final[0..80];
    }
    return final;
}

fn isDigit(c: u8) bool {
    return c >= '0' and c <= '9';
}

fn isHexChar(c: u8) bool {
    return (c >= '0' and c <= '9') or (c >= 'a' and c <= 'f') or (c >= 'A' and c <= 'F');
}
